crates/core/tests/error_notification.rs

```diff
@@ -0,0 +1,187 @@
+//! via the result router, rather than leaving clients hanging indefinitely.
+//!
+//! Related Issues:
+//! - #1858: Clients hang when operations fail (no error notification)
+
+use freenet::{
+    config::{ConfigArgs, NetworkArgs, SecretArgs, WebsocketApiArgs},
+    dev_tool::TransportKeypair,
+    local_node::NodeConfig,
+    server::serve_gateway,
+    test_utils::{load_contract, make_get},
+};
+use freenet_stdlib::{
+    client_api::{ClientRequest, WebApi},
+    prelude::*,
+};
+use futures::FutureExt;
+use std::{
+    net::Ipv4Addr,
+    sync::{LazyLock, Mutex},
+    time::Duration,
+};
+use tokio::{select, time::timeout};
+use tokio_tungstenite::connect_async;
+use tracing::error;
+
+static RNG: LazyLock<Mutex<rand::rngs::StdRng>> = LazyLock::new(|| {
+    use rand::SeedableRng;
+    Mutex::new(rand::rngs::StdRng::from_seed(
+        *b"error_notification_test_seed0123",
+    ))
+});
+
+/// Helper to create a node configuration for testing
+async fn create_test_node_config(
+    is_gateway: bool,
+    ws_api_port: u16,
+    network_port: Option<u16>,
+) -> anyhow::Result<(ConfigArgs, tempfile::TempDir)> {
+    let temp_dir = tempfile::tempdir()?;
+    let key = TransportKeypair::new();
+    let transport_keypair = temp_dir.path().join("private.pem");
+    key.save(&transport_keypair)?;
+    key.public().save(temp_dir.path().join("public.pem"))?;
+
+    let config = ConfigArgs {
+        ws_api: WebsocketApiArgs {
+            address: Some(Ipv4Addr::LOCALHOST.into()),
+            ws_api_port: Some(ws_api_port),
+        },
+        network_api: NetworkArgs {
+            public_address: Some(Ipv4Addr::LOCALHOST.into()),
+            public_port: network_port,
+            is_gateway,
+            skip_load_from_network: true,
+            gateways: Some(vec![]), // Empty gateways for isolated node
+            location: Some({
+                use rand::Rng;
+                RNG.lock().unwrap().random()
+            }),
+            ignore_protocol_checking: true,
+            address: Some(Ipv4Addr::LOCALHOST.into()),
+            network_port,
+            bandwidth_limit: None,
+            blocked_addresses: None,
+        },
+        config_paths: freenet::config::ConfigPathsArgs {
+            config_dir: Some(temp_dir.path().to_path_buf()),
+            data_dir: Some(temp_dir.path().to_path_buf()),
+        },
+        secrets: SecretArgs {
+            transport_keypair: Some(transport_keypair),
+            ..Default::default()
+        },
+        ..Default::default()
+    };
+
+    Ok((config, temp_dir))
+}
+
+/// Test that GET operation errors are delivered to WebSocket clients
+///
+/// This test verifies that when a GET operation fails (e.g., contract not found
+/// on an isolated node), the client receives an error response rather than
+/// hanging indefinitely.
+///
+/// Fixes: #1858
+#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+async fn test_get_error_notification() -> anyhow::Result<()> {
```

Comment:
I see we are only testing one case here, would it be good to have multiple error types tested? 

we can do thing slike connect 2 peers and then drop one of them to force certain edge cases.